### ISSUE TITLE
[mps] [PyTorch 2.0] LayerNorm crashes when input is in `float16`

### ISSUE THREAD

**Issue by pcuenca** [pytorch/pytorch#96113]
### 🐛 Describe the bug

As stated in the title, the following crashes when using the `mps` device:

```python
ln = nn.LayerNorm((768,), elementwise_affine=True).to("mps")
ln(torch.randn(1, 77, 768).to("mps", dtype=torch.float16))
```

### Versions

Collecting environment information...
PyTorch version: 2.1.0.dev20230306
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.2 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: Could not collect
Libc version: N/A

Python version: 3.9.13 | packaged by conda-forge | (main, May 27 2022, 17:01:00)  [Clang 13.0.1 ] (64-bit runtime)
Python platform: macOS-13.2-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Max

Versions of relevant libraries:
[pip3] functorch==0.2.0a0+b2c9d60
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.2
[pip3] torch==2.1.0.dev20230306
[pip3] torchinfo==1.7.0
[conda] numpy                     1.23.2                   pypi_0    pypi
[conda] torch                     2.1.0.dev20230306          pypi_0    pypi
[conda] torchinfo                 1.7.0                    pypi_0    pypi

cc @ezyang @gchanan @zou3519 @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

**Comment by HamidShojanazeri** [comment]
cc: @ngimel

**Comment by malfet** [comment]
Correct me if I'm wrong, but this used to work in 1.13.1, isn't it?

**Comment by malfet** [comment]
Similar problem in `torch.addcmul`:
```
$  python -c "import torch;x=torch.rand(3, 3, dtype=torch.float16, device='mps');y=x.addcmul(torch.ones(3, device='mps'), torch.ones(3, device='mps'));print(y)"
loc("mps_multiply_1"("(mpsFileLoc): /AppleInternal/Library/BuildRoots/9e200cfa-7d96-11ed-886f-a23c4f261b56/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm":228:0)): error: input types 'tensor<3xf32>' and 'tensor<1xf16>' are not broadcast compatible
LLVM ERROR: Failed to infer result type(s).
zsh: abort      python -c 

```

**Comment by kulinseth** [comment]
@skotapati, investigating from MPS side.

**Comment by pcuenca** [comment]
Interesting, I was debugging [this issue](https://github.com/huggingface/diffusers/issues/2521), which is a regression on our side. Since this also happens in PyTorch 1.13.1 (thanks @malfet!), the promotion must be performed somewhere else. Taking another look.

**Comment by kulinseth** [comment]
> @pcuenca , I guess if one wants to use `nn.LayerNorm()` for float16 dtype, one should do something like: `torch.nn.LayerNorm((768,), elementwise_affine=True).to(device="mps",dtype=torch.float16)`, which would be much more efficient that performing LayerNorm with float32 weighs against float16 input tensor

Thanks @malfet . @pcuenca all the normalization layers being BW sensitive, it will be great if numerically the model can support , we perform the computations in similar reduced precision for both inputs and weights. Otherwise we will have to upcast those layers to higher precision and we will lose the BW saving benefits.

**Comment by rituzangle** [comment]
I encountered a similar issue where the `diffusers` library was reporting NSFW content and returning blank images when using `torch.float16`. The error message was:
project_pics/imagr/sd_env/`lib/python3.10/site-packages/diffusers/image_processor.py:147`:` RuntimeWarning: invalid value encountered in cast images = (images * 255).round().astype("uint8") Potential NSFW content was detected in one or more images. A black image will be returned instead. Try again with a different prompt and/or seed.`


Switching the `torch_dtype` from `torch.float16` to `torch.float32` resolved the error, the message about 'NSFW content detected' message as well as the blank / black image, for me. I got pictures. Also, '  safety_checker=None,' didn't get rid of the NSFW message. btw, it flagged a simple image request of "sewing machine on a table" the same way until I changed the torch.float from 16 to 32.

pipeline = StableDiffusionPipeline.from_pretrained(
    "runwayml/stable-diffusion-v1-5",
    torch_dtype=**torch.float32**,
).to(device)

Mac M3, 16GB. Sequoia 15.2
